pkg/minishift/openshift/openshift.go

```diff
@@ -0,0 +1,221 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openshift
+
+import (
+	"errors"
+	"fmt"
+	"github.com/minishift/minishift/pkg/minishift/docker"
+	"github.com/pborman/uuid"
+	"strings"
+	"github.com/golang/glog"
+)
+
+const OPENSHIFT_CONTAINER_NAME = "origin"
+const OPENSHIFT_EXEC = "/usr/bin/openshift"
+
+type OpenShiftPatchTarget struct {
+	target              string
+	containerConfigFile string
+	localConfigFile string
+	ip              string
+}
+
+var (
+	MASTER = OpenShiftPatchTarget{
+		"master",
+		"/var/lib/origin/openshift.local.config/master/master-config.yaml",
+		"/var/lib/minishift/openshift.local.config/master/master-config.yaml",
+		"",
+	}
+
+	NODE = OpenShiftPatchTarget{
+		"node",
+		"/var/lib/origin/openshift.local.config/node-%s/node-config.yaml",
+		"/var/lib/minishift/openshift.local.config/node-%s/node-config.yaml",
+		"",
+	}
+
+	UNKNOWN = OpenShiftPatchTarget{
+		"unkown",
+		"",
+		"",
+		"",
+	}
+)
+
+func (t *OpenShiftPatchTarget) SetIp(ip string) {
+	t.ip = ip
+}
+
+func (t *OpenShiftPatchTarget) containerConfigFilePath() (string, error) {
+	if t.target == "node" {
+		if t.ip == "" {
+			return "", errors.New("Node IP was not set")
+		}
+		return fmt.Sprintf(t.containerConfigFile, t.ip), nil
+	} else {
+		return t.containerConfigFile, nil
+	}
+}
+
+func (t *OpenShiftPatchTarget) localConfigFilePath() (string, error) {
+	if t.target == "node" {
+		if t.ip == "" {
+			return "", errors.New("Node IP was not set")
+		}
+		return fmt.Sprintf(t.localConfigFile, t.ip), nil
+	} else {
+		return t.localConfigFile, nil
+	}
+}
+
+func RestartOpenShift(commander docker.DockerCommander) (bool, error) {
+	fmt.Println("Restarting OpenShift")
+	ok, err := commander.Restart(OPENSHIFT_CONTAINER_NAME)
+	if err != nil {
+		return false, err
+	}
+	return ok, nil
+}
+
+func Patch(target OpenShiftPatchTarget, patch string, commander docker.DockerCommander) (bool, error) {
+	fmt.Println(fmt.Sprintf("Patching OpenShift configuration %s with %s", target.containerConfigFile, patch))
+
+	patchId, err := backUpConfig(target, commander)
+	if err != nil {
+		return false, err
+	}
+
+	containerConfigPath, err := target.containerConfigFilePath()
+	if err != nil {
+		return false, err
+	}
+
+	patchCommand := fmt.Sprintf("ex config patch %s --patch='%s'", containerConfigPath, patch)
+	result, err := commander.Exec("-t", OPENSHIFT_CONTAINER_NAME, OPENSHIFT_EXEC, patchCommand)
+	if err != nil {
+		glog.Error("Creating patched configuration failed. Not applying the changes.", err)
+		return false, nil
+
+	}
+
+	// Tweak the result configuration, we need to escape single quotes
+	result = strings.Replace(result, "'", "'\\''", -1)
+	writeConfig(target, result, commander)
+	if err != nil {
+		rollback(target, commander, patchId)
+		return false, nil
+	}
+
+	_, err = commander.Restart(OPENSHIFT_CONTAINER_NAME)
```

Comment:
Because you want to have the output to the stdout? I guess for the initial attempt that's ok, in the rollback I intentionally call the commander directly, since the output gets confusing otherwise.